### PR TITLE
Realm email setting

### PR DESCRIPTION
**Pull request by whaught**
Issue https://github.com/google/exposure-notifications-verification-server/issues/787

## Proposed Changes

* Per-realm email settings parallel to our SMS settings
    * Encrypt smtp password with our secret, similar to twilio auth token
* Doesn't use the settings for sending the things yet

### CODE REVIEW

File: cmd/server/assets/realmadmin/_form_email.html [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{{define "realmadmin/_form_email"}}
+
+{{$realm := .realm}}
+{{$emailConfig := .smtpConfig}}
+{{$countries := .countries}}
+
+<p class="mb-4">
+  These are the settings for configuring an SMTP email provider. The verification server
+  may use this email account to send invitations, password resets, and account-verifications
```
Comment (sethvargo) [inline cmd/server/assets/realmadmin/_form_email.html:9]: may? Under what conditions would it and wouldn't it?

File: cmd/server/assets/realmadmin/_form_email.html [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    <div class="form-group form-check">
+      <input type="checkbox" name="use_system_smtp_config" id="use-system-smtp-config" class="form-check-input" value="1" {{if $realm.UseSystemSMTPConfig}} checked{{end}}
+        data-toggle="collapse" data-target="#smtp-form">
+      <label class="form-check-label" for="use-system-smtp-config">
+        Use system SMTP configuration
+      </label>
+    </div>
+  {{end}}
+
+  <div id="smtp-form" class="collapse{{if not $realm.UseSystemSMSConfig}} show{{end}}">
```
Comment (sethvargo) [inline cmd/server/assets/realmadmin/_form_email.html:35]: SMS

File: cmd/server/assets/realmadmin/edit.html [hunk 3]
```diff
@@ -39,6 +39,9 @@ <h1>Realm settings</h1>
           <li class="nav-item" role="presentation">
             <a class="nav-link" id="sms-tab" data-toggle="tab" href="#sms" role="tab" aria-controls="sms" aria-selected="false">SMS</a>
           </li>
+          <li class="nav-item" role="presentation">
+            <a class="nav-link" id="sms-email" data-toggle="tab" href="#email" role="tab" aria-controls="email" aria-selected="false">Email</a>
```
Comment (sethvargo) [inline cmd/server/assets/realmadmin/edit.html:43]:
```suggestion
            <a class="nav-link" id="email" data-toggle="tab" href="#email" role="tab" aria-controls="email" aria-selected="false">Email</a>
```

File: pkg/database/email_config.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/jinzhu/gorm"
+)
+
+// EmailConfig represents and email configuration.
+type EmailConfig struct {
+	gorm.Model
+	Errorable
+
+	// email Config belongs to exactly one realm.
+	RealmID uint
```
Comment (sethvargo) [inline pkg/database/email_config.go:31]:
```suggestion
	RealmID uint `gorm:"type:integer"`
```

Comment (sethvargo) [inline pkg/database/email_config.go:31]: Do we need an index or anything here?

File: pkg/database/email_config.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	// underlying configuration.
+	ProviderType email.ProviderType `gorm:"type:varchar(100)"`
+
+	SMTPAccount string `gorm:"type:varchar(250)"`
+	SMTPHost    string `gorm:"type:varchar(250)"`
+	SMTPPort    string `gorm:"type:varchar(250)"`
+
+	// TwilioAuthToken is encrypted/decrypted automatically by callbacks. The
+	// cache fields exist as optimizations.
+	SMTPPassword                string `gorm:"type:varchar(250)"`
```
Comment (sethvargo) [inline pkg/database/email_config.go:43]: Use `text` type instead

File: pkg/database/email_config.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+				Error; err != nil {
+				return err
+			}
+		}
+
+		// The fields were blank, delete the email config.
+		return db.db.Unscoped().Delete(s).Error
+	}
+
+	if db.db.NewRecord(s) {
```
Comment (sethvargo) [inline pkg/database/email_config.go:106]: I think you can just call `db.db.Save(s)`

File: pkg/database/email_config.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		return db.db.Create(s).Error
+	}
+	return db.db.Save(s).Error
+}
+
+// DeleteEmailConfig removes an email configuration record.
+func (db *Database) DeleteEmailConfig(s *EmailConfig) error {
```
Comment (sethvargo) [inline pkg/database/email_config.go:113]: Is this used? The `Save` method seems to handle the deleting.

File: pkg/database/realm.go [hunk 8]
```diff
@@ -142,6 +143,26 @@ type Realm struct {
...
+
+	// EmailPasswordResetTemplate is the template for resetting password.
+	EmailPasswordResetTemplate string `gorm:"type:varchar(400);"`
+
+	// EmailVerifyTemplate is the template used for email verification.
+	EmailVerifyTemplate string `gorm:"type:varchar(400);"`
+
+	// CanUseSystemEmailConfig is configured by system administrators to share the
+	// system email config with this realm. Note that the system email config could be
+	// empty and a local SMS config is preferred over the system value.
```
Comment (sethvargo) [inline pkg/database/realm.go:157]: SMS

File: pkg/database/realm.go [hunk 9]
```diff
@@ -142,6 +143,26 @@ type Realm struct {
...
 	UseSystemSMSConfig bool `gorm:"column:use_system_sms_config; type:bool; not null; default:false;"`
 
+	// EmailInviteTemplate is the template for inviting new users.
+	EmailInviteTemplate string `gorm:"type:varchar(400);"`
+
+	// EmailPasswordResetTemplate is the template for resetting password.
+	EmailPasswordResetTemplate string `gorm:"type:varchar(400);"`
+
+	// EmailVerifyTemplate is the template used for email verification.
+	EmailVerifyTemplate string `gorm:"type:varchar(400);"`
```
Comment (sethvargo) [inline pkg/database/realm.go:153]: text

File: pkg/database/realm.go [hunk 10]
```diff
@@ -142,6 +143,26 @@ type Realm struct {
 	// configuration, making it impossible to opt out of text message sending.
 	UseSystemSMSConfig bool `gorm:"column:use_system_sms_config; type:bool; not null; default:false;"`
 
+	// EmailInviteTemplate is the template for inviting new users.
+	EmailInviteTemplate string `gorm:"type:varchar(400);"`
+
+	// EmailPasswordResetTemplate is the template for resetting password.
+	EmailPasswordResetTemplate string `gorm:"type:varchar(400);"`
```
Comment (sethvargo) [inline pkg/database/realm.go:150]: text

File: pkg/database/realm.go [hunk 11]
```diff
@@ -142,6 +143,26 @@ type Realm struct {
 	// configuration, making it impossible to opt out of text message sending.
 	UseSystemSMSConfig bool `gorm:"column:use_system_sms_config; type:bool; not null; default:false;"`
 
+	// EmailInviteTemplate is the template for inviting new users.
+	EmailInviteTemplate string `gorm:"type:varchar(400);"`
```
Comment (sethvargo) [inline pkg/database/realm.go:147]: text

File: pkg/database/email_config.go (hunk 4 above)
Comment (whaught) [reply pkg/database/email_config.go]: index is added in migrations

**Comment by whaught** [comment]
/close

I've continued working on this in https://github.com/google/exposure-notifications-verification-server/pull/849 it's the sum of this and its uses. It's gotten some good review feedback - let's go for the onme-giant-pr strategy.

**Pull request closed without merging**